substrate/rwlock

```diff
@@ -0,0 +1,49 @@
+// SPDX-License-Identifier: BSD-3-Clause
+#ifndef SUBSTRATE_RWLOCK
+#define SUBSTRATE_RWLOCK
+
+#if __cplusplus < 201703L
+#error "rwlock is on available on C++17 and above"
+#endif
+
+#include <substrate/internal/defs>
+
+#include <utility>
+#include <mutex>
+#include <shared_mutex>
+
+namespace substrate
+{
+	template<typename T>
+	struct rwlock final {
+	private:
+		template<typename U, template<typename> typename lock_t>
+		struct lock_result final {
+		private:
+			lock_t<std::shared_mutex> _lock;
+			U& _obj;
+		public:
+			constexpr lock_result(std::shared_mutex& mut, U& obj) noexcept :
+				_lock{mut}, _obj{obj} { }
+
+			SUBSTRATE_NO_DISCARD(U* operator->() noexcept { return &_obj; })
```

Comment:
Code style dictates this should be `U *operator ->()`, and likewise `U &operator *()` for the next one.

We suggest the closing bracket for the macro is moved back to the noexcept marking to prevent issues should these be changed in the future or the macro revised to append in some cases rather than prepend